platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ResolveExceptionContextualizer.java

```diff
@@ -47,80 +45,83 @@ public ResolveExceptionContextualizer(
-        if (failure instanceof ResolveException) {
-            return (ResolveException) failure;
+        // Keep the original exception if no changes were made to
+        // the causes to avoid losing the original stack trace
+        if (mappedCauses.equals(resolveException.getCauses())) {
+            return resolveException;
         }
 
-        return new DefaultLenientConfiguration.ArtifactResolveException(type, contextDisplayName, Collections.singleton(failure));
+        return new TypedResolveException(resolveException.getType(), contextDisplayName, mappedCauses, resolveException.getResolutions());
     }
 
-    @Nullable
-    private ResolveException detectRepositoryOverride(String contextDisplayName, Collection<? extends Throwable> causes) {
-        try {
-            boolean ignoresSettingsRepositories = false;
-            if (domainObjectContext instanceof ProjectInternal) {
-                ProjectInternal project = (ProjectInternal) domainObjectContext;
-                ignoresSettingsRepositories = !project.getRepositories().isEmpty() &&
-                    !project.getGradle().getSettings().getDependencyResolutionManagement().getRepositories().isEmpty();
-            }
-
-            boolean hasModuleNotFound = causes.stream().anyMatch(ModuleVersionNotFoundException.class::isInstance);
-
-            if (ignoresSettingsRepositories && hasModuleNotFound) {
-                return new ResolveExceptionWithHints(contextDisplayName, causes,
-                    "The project declares repositories, effectively ignoring the repositories you have declared in the settings.\n" +
-                        "You can figure out how project repositories are declared by configuring your build to fail on project repositories.\n" +
-                        documentationRegistry.getDocumentationRecommendationFor("information", "declaring_repositories", "sub:fail_build_on_project_repositories")
-                );
-            }
-
-            return null;
-        } catch (Throwable e) {
-            // To catch `The settings are not yet available for` error
-            return new ResolveException(contextDisplayName, ImmutableList.<Throwable>builder().addAll(causes).add(e).build());
+    // TODO: We should handle this exception at the source instead of using instanceof to detect it after it is thrown.
+    //       We should try to avoid catching and analyzing runtime exceptions
+    public Throwable mapRepositoryOverrideFailure(String contextDisplayName, Throwable failure) {
+        if (!(failure instanceof ModuleVersionNotFoundException) || !settingsRepositoriesIgnored()) {
+            return failure;
         }
+
+        ImmutableList<String> resolutions = ImmutableList.of(
+            "The project declares repositories, effectively ignoring the repositories you have declared in the settings.\n" +
+                "You can figure out how project repositories are declared by configuring your build to fail on project repositories.\n" +
```

Comment:
🤔 This resolution...isn't one.  Sure you could make your build fail in a different way, but how does this address the current runtime failure, whatever it is?

This is useful information, but maybe it should be emitted via a different mechanism.

Reply:
I didn't add this resolution initially, and I agree we should have a better way of adding it at the time of throwing the exception. 

The reason we are adding this to the exception as a resolution is because we have determined that repositories declared at the settings-level are being ignored for this project. We only add this resolution to `ModuleVersionNotFoundException`s. 

We are making the assumption that the module version could not be found because we are ignoring repositories. Perhaps this could be worded better to give the user some actionable way to fix this, but I think this does point the user in the right direction as to _why_ they are failing to resolve a module dependency.

Reply:
> We are making the assumption that the module version could not be found because we are ignoring repositories.

Right, and the error could have nothing to do with this situation.  Can't we just emit a standard warning with the info that project repos are ignored instead?  This resolution could say "Consider defining additional repos in settings.gradle(.kts) to resolve dependency X."

Alternately, if we leave this as a resolution, "You can figure out" is language that would infuriate me - the tool has way more info than I do, _it_ should be the one figuring things out!  Seems like this touches on overall Comprehensibility and UX.

Reply:
Adding new repos to settings would not change the outcome here, as those repos are the ones being ignored. Do you mean we should recommend that the user declare additional _project_ repositories?

As for the "You can figure out", we can use more assertive language here. What about:

`To determine how project repositories are declared, configure your build to fail on project repositories.`

Reply:
When you say "emit a standard warning", you mean with a `Logger.warn`? I feel that decoupling that warning from the error that caused it is a step backwards. 

While we are only assuming that the ignoring of settings repos is the cause of this error, it very well could be the reason. 

This seems to be the original reason this warning was added: https://github.com/gradle/gradle/issues/15336

Reply:
I think we should warn in any case, even without the exception.  A resolution could reference the warning message.

Reply:
Yes, with the logger.

Reply:
The Assertive language suggestions sounds much better to me.  Add to project repos is what I meant - whichever is used.

Reply:
> So, warn even if the build works perfectly fine?

Yes, this is exactly what I'd want to see a warning for - it works now, but might get you into trouble later.

Reply:
If we're going to unconditionally warn in those cases, we should just deprecate being able override settings repos. 

However, I have a feeling there are valid reasons to want to use repos different than the build-wide configured repos, in which case I do not want to see this warning

Reply:
Yeah, I would love to see a `RepositoriesMode.WARN_ON_PROJECT_REPOS` mode added and made the default.  Obviously beyond the scope of this PR.  Warning if you have project repos, at least with `PREFER_SETTINGS`, seems like a good step.

I agree this probably isn't deprecatable, because as you say, it's likely builds use both on purpose for different projects.

I suppose this isn't worth worrying too much about.  At least they see the info after it has become a problem.  Changing the language is a nice positive.

Reply:
Yeah I can change the language here for sure. The existing message says `configure your build to fail on project repositories` but we can actually point to the API that does that instead of relying on the user to find it

Reply:
It actually looks like we already print the relevant documentation URL.